### PR TITLE
Auto Batching impl 

### PR DESCRIPTION
**Pull request by myshevts**

### CODE REVIEW

**Comment by yeonbok** [comment]
FYI, I just wanted to mention about detection output because it seemed that you are going to test autobatch with DetectionOutput. You might know that DetectionOtuput of GPU plugin has both impl for cpu and gpu, and there are some conditons for detection output to select gpu impl : https://github.com/openvinotoolkit/openvino/blob/27fa64031b4f46fdb05ae64ca0f425072808cdd7/inference-engine/thirdparty/clDNN/src/layout_optimizer.cpp#L1150

For batch size, it should be 4<=bsize <= max LWS (256 for DG1, larger # for DG2.. 1024? maybe) . 
If the batch size is not within that range, it will select cpu impl, though the device is GPU. 
Also there are other conditions like topk and the threshold.

**Comment by myshevts** [comment]
>If the batch size is not within that range, it will select cpu impl, though the device is GPU.

thanks for pointing, although for my impl this doesn't make any difference (whether CPU or GPU will execute the DO layer).
In the tests I'm testing that with the batch size of 1 thru 256

File: inference-engine/src/cldnn_engine/cldnn_config.h [hunk 1]
```diff
@@ -14,6 +14,9 @@
 
 namespace CLDNNPlugin {
 
+// fixme: this value should be deduced from the #command-streamers, and presumably queried from the plugin
+const uint32_t default_num_streams_for_tput = 2;
+
 struct Config {
     Config(std::string device_id = "0") : device_id(device_id),
                                           throughput_streams(1),
```
Comment (yeonbok) [inline inference-engine/src/cldnn_engine/cldnn_config.h]: is this to be set as "default_num_streams_for_tput"?

Comment (myshevts) [reply inference-engine/src/cldnn_engine/cldnn_config.h]: by default, the THROUGHPUT mode (that sets the 2) is not enabled, so this is defualt (legacy value)

File: inference-engine/thirdparty/clDNN/api/intel_gpu/runtime/device_info.hpp [hunk 2]
```diff
@@ -25,6 +26,10 @@ struct gfx_version {
     uint16_t major;
     uint8_t minor;
     uint8_t revision;
+    friend bool operator < (const gfx_version& l, const gfx_version& r)  {
```
Comment (ilyachur) [inline inference-engine/thirdparty/clDNN/api/intel_gpu/runtime/device_info.hpp:29]: Are these changes related to PR?

File: cmake/features.cmake [hunk 3]
```diff
@@ -116,6 +116,8 @@ ie_option (ENABLE_GAPI_PREPROCESSING "Enables G-API preprocessing" ON)
 ie_option (ENABLE_MULTI "Enables MULTI Device Plugin" ON)
 ie_option (ENABLE_AUTO "Enables AUTO Device Plugin" ON)
 
+ie_option (ENABLE_BATCH "Enables Auto-Batching Plugin" ON)
```
Comment (ilyachur) [inline cmake/features.cmake:119]: batch is not always associated with plugin, maybe `ENABLE_AUTO_BATCH`?

File: src/inference/dev_api/ie_icore.hpp [hunk 4]
```diff
@@ -60,6 +61,22 @@ class ICore {
...
+     * Users can create as many networks as they need and use
+     *        them simultaneously (up to the limitation of the hardware resources)
+     *
+     * @param network CNNNetwork object acquired from Core::ReadNetwork
+     * @param remoteCtx  "Remote" (non-CPU) accelerator device-specific execution context to use
+     * @param config Optional map of pairs: (config parameter name, config parameter value) relevant only for this load
+     * operation
+     * @return An executable network reference
+     */
+    virtual SoExecutableNetworkInternal LoadNetwork(const CNNNetwork& network,
```
Comment (ilyachur) [inline src/inference/dev_api/ie_icore.hpp:76]: Why do we need this new method?

File: src/inference/include/ie/ie_parallel.hpp [hunk 5]
```diff
@@ -129,6 +133,69 @@ inline void parallel_set_num_threads(int) {
 #endif
 
 namespace InferenceEngine {
+#if ((IE_THREAD == IE_THREAD_TBB) || (IE_THREAD == IE_THREAD_TBB_AUTO))
```
Comment (ilyachur) [inline src/inference/include/ie/ie_parallel.hpp:136]: Should it be a part of this PR?

File: src/inference/src/ie_core.cpp [hunk 6]
```diff
@@ -501,12 +513,132 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
...
+                                       ? deviceNameWithoutBatch + "(" + std::to_string(optimalBatchSize) + ")"
+                                       : deviceNameWithBatchSize;
+                if (bDetectionOutput) {
+                    deviceName = "HETERO:BATCH," + deviceNameWithoutBatch;
+                    config_with_batch[CONFIG_KEY(AUTO_BATCH)] = batchConfig;
+                } else {
+                    deviceName = "BATCH:" + batchConfig;
+                }
+            }
+        } catch (std::exception& e) {
```
Comment (ilyachur) [inline src/inference/src/ie_core.cpp:626]: Should it be `NOT_IMPLEMENTED`?

File: src/inference/src/ie_core.cpp [hunk 7]
```diff
@@ -501,12 +513,132 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
...
+                    optimalBatchSize = std::max(1u, std::min(requests, optimalBatchSize));
+                }
+            }
+            if (optimalBatchSize > 1 || !deviceNameWithBatchSize.empty()) {
+                auto function = network.getFunction();
+                bool bDetectionOutput = false;
+                for (auto&& node : function->get_ops()) {
+                    auto isDetectionOutputParent = [](decltype(node)& nd) {
+                        for (size_t n = 0; n < nd->get_input_size(); n++) {
+                            if (!std::strcmp("DetectionOutput", nd->get_input_node_ptr(n)->get_type_info().name))
```
Comment (ilyachur) [inline src/inference/src/ie_core.cpp:602]: It won't work if we have several versions of DO operations.
Can we use RTTI instead of `strcmp` with `type_info`?

File: src/inference/src/ie_core.cpp [hunk 8]
```diff
@@ -1117,14 +1268,36 @@ std::vector<std::string> DeviceIDParser::getMultiDevices(std::string devicesList
...
+                deviceNames.insert(DeviceIDParser::getBatchDevice(devicesList.substr(p + 1)));
+        } else {
+            deviceNames.insert(trim_request_info(devicesList));
+        }
+    }
+    return std::vector<std::string>(deviceNames.begin(), deviceNames.end());
+}
 
-    return deviceNames;
+std::string DeviceIDParser::getBatchDevice(std::string device) {
```
Comment (ilyachur) [inline src/inference/src/ie_core.cpp:1295]:
```suggestion
std::string DeviceIDParser::getBatchDevice(const std::string& device) {
```

File: src/inference/src/ie_core.cpp [hunk 9]
```diff
@@ -1117,14 +1268,36 @@ std::vector<std::string> DeviceIDParser::getMultiDevices(std::string devicesList
...
+    auto trim_request_info = [](std::string device_with_requests) {
```
Comment (ilyachur) [inline src/inference/src/ie_core.cpp:1296]:
```suggestion
    auto trim_request_info = [](const std::string& device_with_requests) {
```

File: src/inference/src/ie_core.cpp [hunk 10]
```diff
@@ -1105,7 +1256,7 @@ std::vector<std::string> DeviceIDParser::getHeteroDevices(std::string fallbackDe
 }
 
 std::vector<std::string> DeviceIDParser::getMultiDevices(std::string devicesList) {
```
Comment (ilyachur) [inline src/inference/src/ie_core.cpp:1258]: const string&

File: src/plugins/auto_batch/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+file(GLOB SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/*.cpp)
+
+file(GLOB HEADERS ${CMAKE_CURRENT_SOURCE_DIR}/*.hpp)
+
+ie_add_plugin(NAME ${TARGET_NAME}
+              DEVICE_NAME "BATCH"
+              SOURCES ${SOURCES} ${HEADERS}
+              VERSION_DEFINES_FOR auto_batch.cpp)
+
+target_link_libraries(${TARGET_NAME} PRIVATE ngraph inference_engine_transformations Threads::Threads)
```
Comment (ilyachur) [inline src/plugins/auto_batch/CMakeLists.txt:16]: Why do we need threads here?
```suggestion
target_link_libraries(${TARGET_NAME} PRIVATE Threads::Threads)
```

File: src/plugins/auto_batch/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+ie_add_api_validator_post_build_step(TARGET ${TARGET_NAME})
```
Comment (ilyachur) [inline src/plugins/auto_batch/CMakeLists.txt:18]: Can we enable code style check for new plugin?

File: src/plugins/auto_batch/CMakeLists.txt (hunk 11 above)
Comment (ilyachur) [reply src/plugins/auto_batch/CMakeLists.txt:16]: Maybe will be better to use:
```
set_ie_threading_interface_for(${TARGET_NAME})
```

Comment (myshevts) [reply src/plugins/auto_batch/CMakeLists.txt:16]: no, this doesn't work, as the code needs the pthreads (for std::therad), just like the ov_runime :)

File: inference-engine/thirdparty/clDNN/api/intel_gpu/runtime/device_info.hpp (hunk 2 above)
Comment (myshevts) [reply inference-engine/thirdparty/clDNN/api/intel_gpu/runtime/device_info.hpp:29]: to sort by the gfx_version in the [std::map](https://github.com/openvinotoolkit/openvino/pull/7883/files#diff-cbdb092fdcbc1aaddd07cd8cf329b73c35c1834bfa35a9420872989eb12c0e20R732)

File: src/inference/dev_api/ie_icore.hpp (hunk 4 above)
Comment (myshevts) [reply src/inference/dev_api/ie_icore.hpp:76]: yes, the plugins (and Auto-BAtching) use the Core thru the IECore, so the method (LoadNetwork with RemoteConetxt) is not new, it just exposes the Core functionality to plugins:https://github.com/openvinotoolkit/openvino/pull/7883/files#diff-f4dddf68fff38766cf43b5c758e47e1e1f9e3d395dc5cea2048c662935fc82d3R1376
same for CreateContext below

File: cmake/features.cmake (hunk 3 above)
Comment (myshevts) [reply cmake/features.cmake:119]: yep

File: src/inference/src/ie_core.cpp (hunk 6 above)
Comment (myshevts) [reply src/inference/src/ie_core.cpp:626]: No, we don't want to expose ANY failures related to the Auto-Batching (and silently fallback to use the batch1 network on the device directly)

File: src/inference/include/ie/ie_parallel.hpp (hunk 5 above)
Comment (myshevts) [reply src/inference/include/ie/ie_parallel.hpp:136]: good point, will move to some internal new (?) header (common) instead

File: src/inference/src/ie_core.cpp (hunk 8 above)
Comment (myshevts) [reply src/inference/src/ie_core.cpp:1295]: this just follows the legacy code of the core (getMultiDevice, getHeteroDevice, etc), notice that std::string is ok to pass by value (compiler moves the things anyway

File: src/inference/src/ie_core.cpp (hunk 10 above)
Comment (myshevts) [reply src/inference/src/ie_core.cpp:1258]: same as above

File: src/plugins/auto_batch/CMakeLists.txt (hunk 12 above)
Comment (myshevts) [reply src/plugins/auto_batch/CMakeLists.txt:18]: great!

File: src/inference/src/ie_core.cpp [hunk 13]
```diff
@@ -501,12 +514,146 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
...
+            if (batch_mode != config_with_batch.end()) {
+                const auto disabled = batch_mode->second == CONFIG_VALUE(NO);
+                // no need for this config key in the rest of loading
+                config_with_batch.erase(batch_mode);
+                if (disabled)
+                    return;
+            }
+        }
+        auto device = ov::runtime::parseDeviceNameIntoConfig(deviceName);
+        if (device._deviceName.find("GPU") != std::string::npos) {
```
Comment (apankratovantonp) [inline src/inference/src/ie_core.cpp:539]: What if user will use not "GPU" device alias?

File: src/plugins/auto_batch/auto_batch.cpp [hunk 14]
```diff
@@ -0,0 +1,718 @@
...
+    // this executor starts the inference while  the task (checking the result) is passed to the next stage
+    struct ThisRequestExecutor : public ITaskExecutor {
+        explicit ThisRequestExecutor(AutoBatchAsyncInferRequest* _this_) : _this{_this_} {}
+        void run(Task task) override {
+            auto& workerInferRequest = _this->_inferRequest->_myBatchedRequestWrapper;
+            std::pair<AutoBatchAsyncInferRequest*, InferenceEngine::Task> t;
+            t.first = _this;
+            t.second = std::move(task);
+            workerInferRequest._tasks.push(t);
+            // it is ok to call unsafe_size() here as the queue only grows (and the bulk removal happens under the
```
Comment (apankratovantonp) [inline src/plugins/auto_batch/auto_batch.cpp:240]: Unsafe_size() should be called only in a safe context. For example when it is guaranteed that is safe to call this function.

File: src/inference/src/ie_core.cpp [hunk 15]
```diff
@@ -501,12 +514,146 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
...
+                options["MODEL_PTR"] = std::const_pointer_cast<ngraph::Function>(network.getFunction());
+                optimalBatchSize = GetCPPPluginByName(DeviceIDParser(deviceNameWithoutBatch).getDeviceName())
+                                       .get_metric(METRIC_KEY(OPTIMAL_BATCH_SIZE), options)
+                                       .as<unsigned int>();
+                auto res =
+                    GetConfig(deviceNameWithoutBatch, CONFIG_KEY(PERFORMANCE_HINT_NUM_REQUESTS)).as<std::string>();
+                requests = PerfHintsConfig::CheckPerformanceHintRequestValue(res);
+                const auto& reqs = config_with_batch.find(CONFIG_KEY(PERFORMANCE_HINT_NUM_REQUESTS));
+                if (reqs != config_with_batch.end())
+                    requests = (unsigned int)PerfHintsConfig::CheckPerformanceHintRequestValue(reqs->second);
```
Comment (apankratovantonp) [inline src/inference/src/ie_core.cpp:606]: static_cast

File: src/plugins/auto_batch/auto_batch.cpp [hunk 16]
```diff
@@ -0,0 +1,718 @@
...
+                IE_ASSERT(workerRequestPtr->_completionTasks.size() == (size_t)workerRequestPtr->_batchSize);
+                // notify the individual requests on the completion
+                for (int c = 0; c < workerRequestPtr->_batchSize; c++) {
+                    workerRequestPtr->_completionTasks[c]();
+                }
+                // reset the timeout
+                workerRequestPtr->_cond.notify_one();
+            });
+
+        workerRequestPtr->_thread = std::thread([workerRequestPtr, this] {
```
Comment (apankratovantonp) [inline src/plugins/auto_batch/auto_batch.cpp:331]: Can the scheduling be done without a separate thread?

File: src/plugins/auto_batch/auto_batch.cpp [hunk 17]
```diff
@@ -0,0 +1,718 @@
...
+            while (1) {
+                std::unique_lock<std::mutex> lock(workerRequestPtr->_mutex);
+                auto status = workerRequestPtr->_cond.wait_for(lock, std::chrono::milliseconds(_timeOut));
```
Comment (apankratovantonp) [inline src/plugins/auto_batch/auto_batch.cpp:334]: Can we eliminate timeout from scheduling?

File: src/plugins/auto_batch/auto_batch.cpp [hunk 18]
```diff
@@ -0,0 +1,718 @@
...
+                for (int c = 0; c < workerRequestPtr->_batchSize; c++) {
+                    workerRequestPtr->_completionTasks[c]();
+                }
+                // reset the timeout
+                workerRequestPtr->_cond.notify_one();
+            });
+
+        workerRequestPtr->_thread = std::thread([workerRequestPtr, this] {
+            while (1) {
+                std::unique_lock<std::mutex> lock(workerRequestPtr->_mutex);
```
Comment (apankratovantonp) [inline src/plugins/auto_batch/auto_batch.cpp:333]: too many operations under the lock

File: src/plugins/auto_batch/auto_batch.cpp (hunk 16 above)
Comment (myshevts) [reply src/plugins/auto_batch/auto_batch.cpp:331]: thread is needed as there maybe costly operations like inputs/outputs copies

File: src/plugins/auto_batch/auto_batch.cpp (hunk 17 above)
Comment (myshevts) [reply src/plugins/auto_batch/auto_batch.cpp:334]: when the "full" bacth is collected the thread is notified (without waiting for the timeout)

File: src/inference/src/ie_core.cpp (hunk 13 above)
Comment (myshevts) [reply src/inference/src/ie_core.cpp]: fixed, switched to check the OPTIMAL_BATCHS_SIZE metric support

File: src/inference/src/ie_core.cpp (hunk 7 above)
Comment (myshevts) [reply src/inference/src/ie_core.cpp:627]: as we discussed in the chat, the versions (opsets) of the DetectionOutput doesn't matter

File: src/inference/src/ie_core.cpp [hunk 19]
```diff
@@ -505,12 +518,156 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
...
+        if (deviceName.find("BATCH") != std::string::npos) {
+            // explicitly enabled Auto-Batching
+            auto pos = deviceName.find_first_of(":");
+            if (pos != std::string::npos) {
+                deviceNameWithBatchSize = deviceName.substr(pos + 1);
+                deviceNameWithoutBatch = DeviceIDParser::getBatchDevice(deviceNameWithBatchSize);
+            }
+        } else {
+            // check whether  the Auto-Batching is disabled explicitly
+            const auto& batch_mode = config_with_batch.find(CONFIG_KEY(ALLOW_AUTO_BATCHING));
```
Comment (ilya-lavrenov) [inline src/inference/src/ie_core.cpp:534]: what if `ALLOW_AUTO_BATCHING` is set globally as `core.set_config({ ALLOW_AUTO_BATCHING, OFF })`? I suppose it does not work currently.
See how it's implemented for `CACHE_DIR`

File: src/inference/src/ie_core.cpp [hunk 20]
```diff
@@ -505,12 +518,156 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
...
+                    return;
+            }
+            // check whether if the Auto-Batching is applicable to the device
+            deviceNameWithoutBatch = deviceName;
+            auto pluginName = DeviceIDParser(deviceNameWithoutBatch).getDeviceName();
+            std::vector<std::string> metrics =
+                GetCPPPluginByName(pluginName).get_metric(METRIC_KEY(SUPPORTED_METRICS), {});
+            auto it = std::find(metrics.begin(), metrics.end(), METRIC_KEY(OPTIMAL_BATCH_SIZE));
+            if (metrics.end() == it)
+                return;
```
Comment (ilya-lavrenov) [inline src/inference/src/ie_core.cpp:549]: should not we also return if `OPTIMAL_BATCH_SIZE == 1`?

File: src/inference/src/ie_core.cpp [hunk 21]
```diff
@@ -505,12 +518,156 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
...
+                IE_THROW(NotImplemented)
+                    << "Auto-batching supports only networks featuring inputs/outputs with the batched layouts !";
+
+            unsigned int requests = 0;
+            unsigned int optimalBatchSize = 0;
+            if (deviceNameWithBatchSize.empty()) {
+                // batch size is not set explicitly via device name e.g. BATCH:GPU(4)
+                // let's query the optimal batch size
+                std::map<std::string, ie::Parameter> options;
+                options["MODEL_PTR"] = std::const_pointer_cast<ngraph::Function>(network.getFunction());
```
Comment (ilya-lavrenov) [inline src/inference/src/ie_core.cpp:604]: why do we need to `const_pointer_cast`? I suppose plugin can analyze the const function as well

File: src/inference/src/ie_core.cpp [hunk 22]
```diff
@@ -505,12 +518,156 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
...
+                    deviceName = "HETERO:BATCH," + deviceNameWithoutBatch;
+                    config_with_batch[CONFIG_KEY(AUTO_BATCH)] = batchConfig;
+                } else {
+                    deviceName = "BATCH:" + batchConfig;
+                }
+            }
+        } catch (std::exception& e) {
+            // No Auto-Batching enabled
+        }
+    }
```
Comment (ilya-lavrenov) [inline src/inference/src/ie_core.cpp:658]: can we move this function outside of the `ie_core.cpp`. It's 150 lines of bull shit which is not Core related

File: src/inference/src/ie_core.cpp [hunk 23]
```diff
@@ -505,12 +518,156 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
...
+                            // the code below doesn't need to separate the versions (opsets) of the DetectionOutput
+                            // so type_info name check is enough
+                            // (if in a future there will be a new ver that doesn't mix the batch, this will be new op)
+                            if (!std::strcmp("DetectionOutput", nd->get_input_node_ptr(n)->get_type_info().name))
+                                return true;
+                        }
+                        return false;
+                    };
+
+                    if (!std::strcmp("DetectionOutput", node->get_type_info().name) ||
```
Comment (ilya-lavrenov) [inline src/inference/src/ie_core.cpp:637]: let's compare using `ngraph::op::DetectionOutput::get_type_info_static()`

File: src/plugins/auto/executable_network.cpp [hunk 24]
```diff
@@ -280,6 +281,11 @@ void MultiDeviceExecutableNetwork::TryToLoadNetWork(AutoLoadContext& context,
     auto& deviceList = context.metaDevices;
     bool curDevIsCPU = (device.find("CPU") != std::string::npos);
     try {
+        const auto& mode = deviceConfig.find(CONFIG_KEY(PERFORMANCE_HINT));
+        if (mode !=  deviceConfig.end() && mode->second == CONFIG_VALUE(THROUGHPUT)
+                && device.find("GPU") != std::string::npos)
```
[the remark this answers is not in the document]
Comment (myshevts) [reply src/plugins/auto/executable_network.cpp:286]: the AUTO and the Auto-BAtching a separate story with the dedicated PR WIP (that was a partial merge of that to test,
NB: for @wangleis @tiger100256-hu and @songbell - you should be checking that the ALLO_AUTO_BATCHING is not set to the NO, before pushing YES)

File: src/inference/src/ie_core.cpp [hunk 25]
```diff
@@ -480,14 +485,22 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
...
         if (context == nullptr) {
             IE_THROW() << "Remote context is null";
         }
+        // have to deduce the device name/config from the context first
         auto parsed = parseDeviceNameIntoConfig(context->getDeviceName(), config);
+        std::string& deviceName = parsed._deviceName;
+        std::map<std::string, std::string>& config_with_batch = parsed._config;
+        // if auto-batching is applicable, the below function will patch the device name and config accordingly:
+        ApplyAutoBatching(network, deviceName, config_with_batch);
+        parsed = parseDeviceNameIntoConfig(deviceName, config_with_batch);
```
[the remark this answers is not in the document]
Comment (myshevts) [reply src/inference/src/ie_core.cpp:502]: yes, this is handled in the Auto-BActhing: https://github.com/openvinotoolkit/openvino/pull/7883/files#diff-d004c7047eb69283807e51b6ff1ef834b76a5b44339b2679d8caa9315d07cdc2R683

File: src/inference/src/ie_core.cpp (hunk 21 above)
Comment (myshevts) [reply src/inference/src/ie_core.cpp:604]: this is to mimic the behavior of the MAX_BATCH_SIZE, which had a heated discussion, and this is general limitation of ov::Parameter that it seperates const vs non const ( so you will have a cast in one case anyway)

File: src/inference/src/ie_core.cpp (hunk 20 above)
Comment (myshevts) [reply src/inference/src/ie_core.cpp:549]: pls see line 620 below

File: src/inference/src/ie_core.cpp (hunk 22 above)
Comment (apankratovantonp) [reply src/inference/src/ie_core.cpp:658]: Move to plugin load network itself. You can load and return the network from any plugin even inside the auto batch plugin as you have access to ICore API inside the plugin.

File: src/inference/src/ie_core.cpp (hunk 23 above)
Comment (myshevts) [reply src/inference/src/ie_core.cpp]: fixed

File: src/inference/include/ie/ie_plugin_config.hpp [hunk 26]
```diff
@@ -243,13 +244,26 @@ DECLARE_CONFIG_VALUE(THROUGHPUT);
...
 DECLARE_CONFIG_VALUE(NO);
 
+/**
+ * @brief Auto-batching configuration, string for the device + batch size, e.g. "GPU(4)"
+ */
+DECLARE_CONFIG_KEY(AUTO_BATCH);
+/**
+ * @brief Auto-batching configuration: string with timeout (in ms), e.g. "100"
+ */
+DECLARE_CONFIG_KEY(AUTO_BATCH_TIMEOUT);
```
[the remark this answers is not in the document]
Comment (myshevts) [reply src/inference/include/ie/ie_plugin_config.hpp:276]: good point, will fix separately as part of the CVS-74713

File: src/inference/src/ie_core.cpp (hunk 22 above)
Comment (myshevts) [reply src/inference/src/ie_core.cpp]: will completely refactor later, now simply remove to disable the Auto-batching as being on by default in the product

File: src/inference/src/ie_core.cpp (hunk 19 above)
Comment (myshevts) [reply src/inference/src/ie_core.cpp]: will be fixed later as part of the CVS-74713

File: src/plugins/auto_batch/auto_batch.cpp [hunk 27]
```diff
@@ -0,0 +1,731 @@
...
+                IE_ASSERT(workerRequestPtr->_completionTasks.size() == (size_t)workerRequestPtr->_batchSize);
+                // notify the individual requests on the completion
+                for (int c = 0; c < workerRequestPtr->_batchSize; c++) {
+                    workerRequestPtr->_completionTasks[c]();
+                }
+                // reset the timeout
+                workerRequestPtr->_cond.notify_one();
+            });
+
+        workerRequestPtr->_thread = std::thread([workerRequestPtr, this] {
```
Comment (apankratovantonp) [inline src/plugins/auto_batch/auto_batch.cpp:330]: It is better to use our executor manager and cached thread to eliminate thread recreation.

**Review by apankratovantonp — APPROVED** [review]
Approved.  But:
- It is needed to remove additional thread and timeout based scheduling
- Move all plugin logic into plugin from core
- Remove all logic that is bound to predefine plugins

**Comment by myshevts** [comment]
> It is needed to remove additional thread and timeout based scheduling

Let's discuss offline but timeout is needed to avoid hang in the serial exection (via Infer calls), in the end of app (no full batch collected)

> Move all plugin logic into plugin from core

Currently the only logic in the core relates to the codepath with the HETERO

> Remove all logic that is bound to predefine plugins

This is also already was done (so for the implicit/throughput-hint code path which is now disabled, the impl queried the OPTIMAL_BACTH_SIZE of any device and if it is > 1 for the network, the auto-batching was triggered

File: src/plugins/auto_batch/auto_batch.cpp (hunk 27 above)
Comment (myshevts) [reply src/plugins/auto_batch/auto_batch.cpp:330]: good point - to minimize the mem fragmentation, will fix in the brushing "PR"

**Review by vladimir-paramuzov — APPROVED** [review]
